case-lib/lib.sh

```diff
@@ -1,43 +1,70 @@
 #!/bin/bash
 
+# get test-case information
+SCRIPT_HOME="$(dirname "$0")"
+# get test-case parent folder name
+SCRIPT_HOME=$(cd "$SCRIPT_HOME/.." && pwd)
+# get test-case script name
+SCRIPT_NAME="$0"
+# get test-case parameter
+SCRIPT_PRAM="$*"
+export SCRIPT_HOME SCRIPT_NAME SCRIPT_PRAM
+
 # Source from the relative path of current folder
-source $(dirname ${BASH_SOURCE[0]})/config.sh
-source $(dirname ${BASH_SOURCE[0]})/opt.sh
-source $(dirname ${BASH_SOURCE[0]})/logging_ctl.sh
-source $(dirname ${BASH_SOURCE[0]})/pipeline.sh
-source $(dirname ${BASH_SOURCE[0]})/hijack.sh
+# shellcheck disable=SC1091 source=./config.sh
+source "$SCRIPT_HOME/case-lib/config.sh"
+# shellcheck disable=SC1091 source=./opt.sh
+source "$SCRIPT_HOME/case-lib/opt.sh"
+# shellcheck disable=SC1091 source=./logging_ctl.sh
+source "$SCRIPT_HOME/case-lib/logging_ctl.sh"
+# shellcheck disable=SC1091 source=./pipeline.sh
+source "$SCRIPT_HOME/case-lib/pipeline.sh"
+# shellcheck disable=SC1091 source=./hijack.sh
+source "$SCRIPT_HOME/case-lib/hijack.sh"
+
+# restrict bash version for some bash feature
+[[ $(( BASH_VERSINFO[0] * 10 + BASH_VERSINFO[1] )) -lt 41 ]] && {
+    dlogw "Bash version: ${BASH_VERSINFO[0]}.${BASH_VERSINFO[1]} should > 4.1"
+    exit 2
+}
 
-# force ask buffer data write into file system
+# force ask buffer data write into file system	# force ask buffer data write into file system
 sudo sync -f
 
 # Add tools to command PATH
-if [ ! "$(declare -p TOOL_PATH 2>/dev/null)" ]; then
-    declare -x TOOL_PATH=$(realpath $(dirname ${BASH_SOURCE[1]})/../tools)
-    PATH=$TOOL_PATH:$PATH
+# this line equal `! $(echo $PATH|grep "$SCRIPT_HOME/tools")`
+if [[ ! $PATH =~ $SCRIPT_HOME/tools ]]; then
+    export PATH=$SCRIPT_HOME/tools:$PATH
 fi
 
 # setup SOFCARD id
 if [ ! "$SOFCARD" ]; then
     SOFCARD=$(grep '\]: sof-[a-z]' /proc/asound/cards|awk '{print $1;}')
 fi
+export SOFCARD
 
 if [ ! "$DMESG_LOG_START_LINE" ]; then
-    declare -g DMESG_LOG_START_LINE=$(wc -l /var/log/kern.log|awk '{print $1;}')
+    DMESG_LOG_START_LINE=$(wc -l /var/log/kern.log|awk '{print $1;}')
 fi
+export DMESG_LOG_START_LINE
 
-declare -g SOF_LOG_COLLECT=0
+export SOF_LOG_COLLECT=0
 
 func_lib_setup_kernel_last_line()
 {
-    declare -g KERNEL_LAST_LINE=$(wc -l /var/log/kern.log|awk '{print $1;}')
+    KERNEL_LAST_LINE=$(wc -l /var/log/kern.log|awk '{print $1;}')
+    export KERNEL_LAST_LINE 
```

Comment:
See above. `declare -g` is a no-op, drop it (`export` is `declare -x`, not `declare -g`)